File: android/app/src/main/java/com/jxcode/android/data/ModelCatalog.kt
package com.jxcode.android.data

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import java.net.HttpURLConnection
import java.net.URL

/**
 * Fetches a backend's model list.
 *
 * Four envelope shapes are handled because in practice they all occur:
 *
 *   {"data":[{"id":"..."}]}   OpenAI, vLLM, llama-server
 *   {"data":["..."]}          several gateways
 *   ["..."]                   Ollama's non-standard route
 *   {"models":[...]}          llama-server's older build
 *
 * A backend that answers something else is reported with its raw shape rather
 * than silently yielding an empty list, which the UI would render as "this
 * provider has no models".
 */
object ModelCatalog {

    private const val TIMEOUT_MS = 20_000
    private val json = Json { ignoreUnknownKeys = true; isLenient = true }

    data class Fetch(
        val ids: List<String>,
        val note: String? = null,
        val error: String? = null
    ) {
        val isEmpty get() = ids.isEmpty()
    }

    fun modelsURL(baseURL: String): String {
        val base = baseURL.trim().trimEnd('/')
        return when {
            base.endsWith("/models") -> base
            base.endsWith("/v1") -> "$base/models"
            base.endsWith("/v1/") -> "${base}models"
            else -> "$base/v1/models"
        }
    }

    suspend fun fetch(provider: Provider, apiKey: String?): Fetch =
        withContext(Dispatchers.IO) {
            val url = modelsURL(provider.normalizedBaseURL)
            var connection: HttpURLConnection? = null
            try {
                connection = (URL(url).openConnection() as HttpURLConnection).apply {
                    requestMethod = "GET"
                    connectTimeout = TIMEOUT_MS
                    readTimeout = TIMEOUT_MS
                    setRequestProperty("Accept", "application/json")
                    if (!apiKey.isNullOrBlank()) {
                        // Anthropic's own header and the OpenAI one: sending
                        // both costs nothing and covers either shape.
                        setRequestProperty("x-api-key", apiKey)
                        setRequestProperty("Authorization", "Bearer $apiKey")
                    }
                }
                val status = connection.responseCode
                if (status !in 200..299) {
                    return@withContext Fetch(emptyList(), error = "HTTP $status from $url")
                }
                val text = connection.inputStream.bufferedReader().readText()
                val ids = parse(text)
                if (ids.isEmpty()) {
                    Fetch(emptyList(), note = "no model ids recognised in the response from $url")
                } else {
                    Fetch(ids)
                }
            } catch (e: Exception) {
                Fetch(emptyList(), error = e.message ?: e.toString())
            } finally {
                connection?.disconnect()
            }
        }

    fun parse(text: String): List<String> {
        val element = try {
            json.parseToJsonElement(text)
        } catch (_: Throwable) {
            return emptyList()
        }
        return when (element) {
            is JsonArray -> element.mapNotNull { it.idOrName() }
            is JsonObject -> {
                val data = element["data"]
                if (data != null) {
                    when (data) {
                        is JsonArray -> data.mapNotNull { it.idOrName() }
                        else -> emptyList()
                    }
                } else {
                    val models = element["models"]
                    when (models) {
                        is JsonArray -> models.mapNotNull { it.idOrName() }
                        is JsonObject -> models["data"]?.let { (it as? JsonArray)?.mapNotNull { e -> e.idOrName() } }
                            ?: emptyList()
                        else -> emptyList()
                    }
                }
            }
            else -> emptyList()
        }.distinct()
    }

    private fun JsonElement.idOrName(): String? = when (this) {
        is JsonPrimitive -> contentOrNull?.takeIf { it.isNotBlank() }
        is JsonObject -> (this["id"] ?: this["name"] ?: this["model"])
            ?.let { (it as? JsonPrimitive)?.contentOrNull }
            ?.takeIf { it.isNotBlank() }
        else -> null
    }
}

File: Sources/JXCodeCore/ModelCatalog.swift
import Foundation

/// What a probe of a registered backend found.
public struct ProbeOutcome: Sendable {
    public var models: [String]
    /// Human-readable observations worth showing in the UI — which shape the
    /// server answered in, whether the base URL had to be adjusted, and so on.
    public var notes: [String]

    public init(models: [String], notes: [String] = []) {
        self.models = models
        self.notes = notes
    }
}

/// Fetches model lists from registered backends.
///
/// Four shapes have to be handled, and guessing wrong is the single most common
/// reason a self-hosted provider "doesn't work":
///
///  - `{data:[{id}]}` — OpenAI, vLLM, LM Studio, llama-server, OpenRouter,
///    Together, and Anthropic's own `/v1/models`.
///  - `{models:[{name}]}` — Ollama's native API.
///  - `{data:[{id}]}` at the *root* rather than under `/v1` — llama-server
///    again, depending on version and flags.
///  - A bare JSON array of strings — some small self-hosted gateways.
///
/// The last three are why `fetchModels` falls back rather than failing on the
/// first shape mismatch.
public struct ModelCatalog: ModelCatalogFetching {

    private let session: URLSession
    private let timeout: TimeInterval

    public init(timeout: TimeInterval = 15) {
        let configuration = URLSessionConfiguration.ephemeral
        configuration.timeoutIntervalForRequest = timeout
        configuration.timeoutIntervalForResource = timeout
        // A model list is worth re-fetching; never let a cached one hide a
        // server that has been restarted with different models loaded.
        configuration.requestCachePolicy = .reloadIgnoringLocalCacheData
        self.session = URLSession(configuration: configuration)
        self.timeout = timeout
    }

    // MARK: - Public API

    public func fetchModels(from provider: Provider) async throws -> [String] {
        try await probe(provider).models
    }

    /// Fetch the model list and report how the server answered.
    public func probe(_ provider: Provider) async throws -> ProbeOutcome {
        var notes: [String] = []

        let normalized = provider.normalizedBaseURL
        if normalized != provider.baseURL.trimmingCharacters(in: CharacterSet(charactersIn: "/")) {
            notes.append("base URL normalised to \(normalized)")
        }

        guard let primary = provider.modelsURL else {
            throw ModelCatalogError.invalidURL(provider.baseURL)
        }

        let (data, response) = try await get(primary, provider: provider)

        // A 404 on the primary path is usually a wrong prefix rather than a
        // wrong host, so try the root form before giving up. This is what makes
        // an older llama-server build work without the user knowing why.
        if response.statusCode == 404, provider.kind.acceptsAutomaticV1Prefix {
            if let rootURL = rootModelsURL(for: provider) {
                let (rootData, rootResponse) = try await get(rootURL, provider: provider)
                if (200..<300).contains(rootResponse.statusCode) {
                    let models = try decodeModels(rootData, provider: provider)
                    if !models.isEmpty {
                        notes.append("models found at \(rootURL.path) — this server does not use a /v1 prefix")
                        return ProbeOutcome(models: models, notes: notes)
                    }
                }
            }
        }

        guard (200..<300).contains(response.statusCode) else {
            throw ModelCatalogError.httpStatus(
                response.statusCode,
                String(decoding: data, as: UTF8.self)
            )
        }

        let models = try decodeModels(data, provider: provider)
        guard !models.isEmpty else {
            throw ModelCatalogError.emptyCatalog(normalized)
        }

        notes.append("\(models.count) model(s) from \(primary.path)")
        return ProbeOutcome(models: models, notes: notes)
    }

    // MARK: - Requests

    private func get(_ url: URL, provider: Provider) async throws -> (Data, HTTPURLResponse) {
        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        request.timeoutInterval = timeout
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        for (field, value) in provider.kind.authHeaders(apiKey: provider.apiKey) {
            request.setValue(value, forHTTPHeaderField: field)
        }

        do {
            let (data, response) = try await session.data(for: request)
            guard let http = response as? HTTPURLResponse else {
                throw ModelCatalogError.decoding("response was not HTTP")
            }
            return (data, http)
        } catch let error as ModelCatalogError {
            throw error
        } catch {
            throw ModelCatalogError.transport(error.localizedDescription)
        }
    }

    /// The model list at the server root, ignoring the `/v1` convention.
    private func rootModelsURL(for provider: Provider) -> URL? {
        var base = provider.normalizedBaseURL
        if base.hasSuffix("/v1") { base.removeLast(3) }
        while base.hasSuffix("/") { base.removeLast() }
        return URL(string: base + "/models")
    }

    // MARK: - Decoding

    /// Try each known envelope in turn.
    ///
    /// Order matters only for speed; the shapes are mutually exclusive in
    /// practice, since one has a `data` key, one has `models`, and one is a bare
    /// array.
    ///
    /// A recognised envelope is returned even when it is *empty*, which is what
    /// makes `ModelCatalogError.emptyCatalog` reachable. Previously every branch
    /// required a non-empty result and the function threw on the way out, so an
    /// empty catalog was impossible to report: a server that was running fine
    /// with nothing loaded — the most ordinary state a local backend is ever in
    /// — produced a `decoding` failure whose message was the raw JSON body
    /// instead of "no models loaded".
    private func decodeModels(_ data: Data, provider: Provider) throws -> [String] {
        let decoder = JSONDecoder()

        if let list = try? decoder.decode(OpenAIModelList.self, from: data) {
            return sorted(list.data.map(\.id).filter { !$0.isEmpty })
        }

        if let list = try? decoder.decode(OllamaTagList.self, from: data) {
            return sorted(list.models.map(\.name).filter { !$0.isEmpty })
        }

        // A bare array of strings, which some minimal gateways return.
        if let names = try? decoder.decode([String].self, from: data) {
            return sorted(names.filter { !$0.isEmpty })
        }

        // Last resort: pull every `id` or `name` out of whatever structure this
        // is. Better than failing outright on a server that is otherwise usable
        // — but if even this finds nothing, report a decode failure rather than
        // an empty catalog, because "there are no models" would be a guess.
        if let value = try? decoder.decode(JSONValue.self, from: data) {
            let harvested = harvestIdentifiers(value)
            if !harvested.isEmpty {
                return sorted(harvested)
            }
        }

        let preview = String(decoding: data.prefix(200), as: UTF8.self)
        throw ModelCatalogError.decoding(preview.isEmpty ? "(empty body)" : preview)
    }

    /// Walk an arbitrary JSON tree collecting `id` / `name` string values.
    private func harvestIdentifiers(_ value: JSONValue) -> [String] {
        switch value {
        case .object(let object):
            // Only accept the well-known keys at the object level, so a model's
            // own metadata field called `name` is not mistaken for an id.
            if let id = object["id"]?.stringValue, !id.isEmpty { return [id] }
            if let name = object["name"]?.stringValue, !name.isEmpty { return [name] }

            var found: [String] = []
            for key in ["data", "models", "model", "result", "items"] {
                if let nested = object[key] {
                    found.append(contentsOf: harvestIdentifiers(nested))
                }
            }
            return found

        case .array(let values):
            return values.flatMap { harvestIdentifiers($0) }

        case .string(let value):
            return value.isEmpty ? [] : [value]

        default:
            return []
        }
    }

    /// Alphabetical, case-insensitive, de-duplicated.
    ///
    /// Stable ordering matters because the model picker keeps its selection by
    /// index in some views, and a server that returns its list in a different
    /// order each call would make the selection jump.
    private func sorted(_ names: [String]) -> [String] {
        var seen = Set<String>()
        return names
            .filter { seen.insert($0).inserted }
            .sorted { $0.localizedCaseInsensitiveCompare($1) == .orderedAscending }
    }
}

// MARK: - Model name helpers

public extension Provider {
    /// Whether a model looks like it can do tool calling.
    ///
    /// Used to warn in the UI rather than to block anything. Base models and
    /// most non-instruct models cannot, and an agent pointed at one produces a
    /// confusing loop of plain text where tool calls were expected.
    static func looksToolCapable(_ model: String) -> Bool {
        let lowered = model.lowercased()
        let disqualifying = ["embed", "rerank", "whisper", "tts", "vision-only", "base"]
        return !disqualifying.contains { lowered.contains($0) }
    }
}
